### ISSUE TITLE
BOS Scale Enhancement

### ISSUE THREAD

**Issue by jsollom-hpe** [Cray-HPE/community#6]
## Abstract

The Boot Orchestration Service (BOS) extends full system boot times unacceptably because it cannot deal efficiently with individual node failures.  It naively expects all nodes within a group to succeed or fail together and may re-run the boot/reboot/etc.. activity even if only one of many nodes has experienced a failure.

We propose to enhance the way the Boot Orchestration Agent handles individual nodes in parallel and address failures individually.  The expected result of this enhancement is that boot times will decrease and BOS will handle failures more efficiently.

## Detailed Problem Statement

BOA applies the operation to all of the nodes in the group at the same time. It expects all of the nodes in a given phase to succeed before it moves on to a subsequent phase. Take a reboot operation as an example. It has three phases: shutdown, power on, and configuration. Prior to Shasta 1.4, if a node failed to execute any phase, BOA would consider the whole phase a failure. The Kubernetes BOA job would exit with a failure, and Kubernetes would restart it from scratch. It would attempt to power down all of the nodes again, even though it had already powered down some of them. This is inefficient because it is applying the same operation to nodes where that operation had already succeeded. This gets worse if the failure occurs in a later phase. Say that BOA is asked to reboot one hundred nodes. It successfully powers down all 100. Then, it attempts to boot them all. 99 nodes boot successfully, but they wait for the 100th node to boot. It will not boot because it has some hardware issue. At that point, BOA would exit with a failure. The restarted BOA again starts from scratch. It powers down all 100 nodes even though 99 had successfully booted. Thus, it is inefficient and wasteful because it is repeating operations that already succeeded. Worse, it gives the nodes that had previously succeeded an opportunity to fail.

Recognizing the deficiencies in this design, BOA was changed in the Shasta 1.4 release to leave any nodes that failed during a phase behind. This meant that BOA was not restarting from scratch to deal with failures. It was simply moving ahead. At the end of its life, BOA would log the nodes that failed along with a command that would invoke a new BOS session that re-attempted the operation on the failed nodes. It was up to the system administrator to issue this new command.

## Internal References

* [CASMCMS-410](https://connect.us.cray.com/jira/browse/CASMCMS-4101) BOS/BOA tunable timeout for booting nodes to move to Ready state
* [CASMCMS-5176](https://connect.us.cray.com/jira/browse/CASMCMS-5176) BOA delay between graceful powerdown request to CAPMC and hard poweroff request  too short
* [CAST-23940](https://connect.us.cray.com/jira/browse/CAST-23940) Can't reboot node (x3000c0s37b0n0)
* [CAST-24484](https://connect.us.cray.com/jira/browse/CAST-24484) Nodes sometimes don't become ready

## External References

The proposed solution was modeled off of the Configuration Framework Service (CFS). It uses a database to keep track of the configuration state of every compute node, both actual state and desired state. It has an operator that configures nodes to make the actual state match the desired state.

* [CFS Design Document](https://connect.us.cray.com/confluence/display/CASM/Shasta+Configuration+Management+Design)
* [cfs-operator](https://stash.us.cray.com/projects/SCMS/repos/cfs-operator/browse)
* [cfs-hwsync-agent](https://stash.us.cray.com/projects/SCMS/repos/cfs-hwsync-agent/browse)
* [cfs-state-reporter](https://stash.us.cray.com/projects/SCMS/repos/cfs-state-reporter/browse)

## Proposed Solution(s)

The proposed solution moves away from having BOS launch individual, short-lived BOAs to handle the work. Instead, records of the desired state for each node will be written to a BOS database. The multiple functions that are wrapped up in BOA will be split into individual, long-running BOS operators that will handle transitioning the nodes to the desired state. There would be a shutdown operator, a power-on operator, etc. The operators can work on each node independently as well as concurrently. There is no worry about a single node failure impacting a BOA job because the BOA job is not handling the operations that are affecting the node, the operators are.

 

Here is a link to the [in-depth design document](https://connect.us.cray.com/confluence/pages/viewpage.action?pageId=190450668)

### Alternate Approach

We attempted to retain the current design and work within the constraints of BOA rather than separating out into individual operators because individual operators is a much larger re-architecture. We used queues maintained with BOA to represent the individual phase and moved nodes the nodes from queue to queue as they progressed. This work quickly became complex and inelegant. It was shelved while we pursued the proposed solution.

### What if we don't solve this problem at this point?

If we do not handle nodes independently, then boot times will suffer. Nodes boot at different rates especially River nodes specifically Gigabyte nodes versus Mountain nodes. If nodes can progress from phase to phase independently, they will return to service more quickly positively impacting boot performance.

If we do not solve this problem now, then BOS will continue to leave failed nodes behind. This adds an additional burden on system administration to follow up to remedy these failed nodes by launching another BOS session. As system size increases the number of node failures will also increase.

### What impact is there beyond the problem statement if we fix the problem now?

The proposed record keeping would likely be an improvement over the status provided by BOA currently. It would help with boot timing profiling as every node would have an individual record.

Removing BOA either in whole or in part will help with cleaning up after BOS. A BOA is a Kubernetes job that has a Kubernetes configMap associated with it. Presently, BOS does not clean up either the BOA job or the configMap. Moving to this new model would remove the need to do this clean-up.

## Further Information

The design proposal envisions a daemon running on the Compute Node which is similar to, but distinct from the cfs-state-reporter.  With this daemon in place, a properly authenticated client can send the boot information for the Compute Node it is running on.
